java-runtime-builder-app/src/main/java/com/google/cloud/runtimes/builder/config/domain/BuildContext.java

```diff
@@ -123,13 +133,53 @@ public void writeDockerFiles() throws IOException {
     try (BufferedWriter writer = Files.newBufferedWriter(dockerFilePath)) {
       writer.write(dockerfile.toString());
     }
+  }
+
+  private void writeDockerIgnore() throws IOException {
+    // If there's nothing to write, return
+    if (dockerignore.getLines().size() < 1) {
+      logger.debug("Skipping .dockerignore generation - nothing to write!");
+      return;
+    }
 
-    // write .dockerignore file, appending to an existing file if it exists
     Path dockerIgnorePath = workspaceDir.resolve(DOCKERIGNORE_NAME);
+    Set<String> existingDockerignoreLines = new HashSet<>();
```

Comment:
Have you considered using [`LinkedHashSet`](https://docs.oracle.com/javase/8/docs/api/java/util/LinkedHashSet.html) and then just rewriting the whole file from scratch? I think this might simplify the logic.